src/vcf/validator.cpp

```diff
@@ -132,12 +132,66 @@ namespace ebi
         }
     }
 
+    void compressed_file_warning(std::string const & file_extension)
+    {
+        BOOST_LOG_TRIVIAL(warning) << "Input file should not be compressed (detected " << file_extension
+            << " compression)";
+    }
+
+    bool compare_extension(std::istream &input,
+                           const std::string &source)
+    {
+        boost::filesystem::path source_name(source);
+        std::string file_extension = source_name.extension().string();
+
+        if (file_extension == RAR || file_extension == TAR || file_extension == TAR_GZ || file_extension == TAR_XZ ||
+            file_extension == TAR_Z || file_extension == ZIP) {
+            compressed_file_warning(file_extension);
+            return true;
+        }
+        return false;
+    }
+
+    bool compare_magic_num(std::istream &input)
+    {
+        unsigned char magic[9];
+        input.read((char*)magic, sizeof(magic));
+        input.seekg(0);
+
+        std::vector<std::pair<std::vector<unsigned char>, std::string>> types = {
+            { { 80, 75, 3, 4 }, ZIP },
+            { { 31, 139 }, TAR_GZ },
+            { { 253, 55, 122, 88, 90 }, TAR_XZ },
+            { { 31, 157}, TAR_Z }
+        };
+
+        for (auto & type : types) {
+            if (std::equal(type.first.begin(), type.first.end(), magic)) {
+                compressed_file_warning(type.second);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    bool is_compressed_file(std::istream &input,
+                            const std::string &source)
+    {
+        if (source != ebi::vcf::STDIN && compare_extension(input, source)) {
```

Comment:
another style thing, put the most general functions first, and right below them the functions they use. e.g.:
```
f1() { f2(); f3();}
f2() { }
f3() { } 
f4() { f5() }
f5() { }
``` 
in this case the order should be `is_compressed_file`, `compare_extension`, `compressed_file_warning`,  `compare_magic_num`.

also, I think changing `compare_extension` to `is_compressed_extension` and `compare_magic_num` to `is_compressed_magic_num` express better what the function returns.

Reply:
what's more, all these functions should be right below is_valid_vcf_file.

Reply:
According to this, I think `compressed_file_warning` should come below `compare_magic_num`.

Reply:
well, it would be with breadth first graph navigation, but we do deep first. In other words, put the used functions after the first usage